handler/src/test/java/io/netty/handler/ssl/ocsp/OcspTest.java

```diff
@@ -0,0 +1,550 @@
+
+    /**
+     * The server has no OCSP staple and handles it like an error. It seems strange
+     * but it's the only strategy for certificates that have the "OCSP Must-Staple"
+     * bit set.
+     *
+     * https://www.grc.com/revocation/ocsp-must-staple.htm
+     * https://blog.mozilla.org/security/2015/11/23/improving-revocation-ocsp-must-staple-and-short-lived-certificates/
+     */
+    private void testServerNonOptionalStaple(SslProvider sslProvider) throws Exception {
+        final AtomicReference<Throwable> clientCauseRef = new AtomicReference<Throwable>();
+        final AtomicReference<Throwable> serverCauseRef = new AtomicReference<Throwable>();
+        final CountDownLatch latch = new CountDownLatch(2);
+
+        ChannelInboundHandlerAdapter serverHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                try {
+                    serverCauseRef.set(cause);
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        ChannelInboundHandlerAdapter clientHandler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                try {
+                    clientCauseRef.set(cause);
+                } finally {
+                    latch.countDown();
+                }
+            }
+        };
+
+        ServerOcspContext serverOcspContext = new DefaultServerOcspContext(null, false);
+        ClientOcspContext clientOcspContext = new TestClientOcspContext(false);
+
+        handshake(sslProvider, latch, serverOcspContext, clientOcspContext, serverHandler, clientHandler);
+        
+        // If the client rejects the staple then Client and Server will fail with
+        // Exceptions which contain these respective messages.
+        String clientMessage = clientCauseRef.get().getMessage();
+        assertTrue(clientMessage, clientMessage.contains("error:1412F113:SSL routines:ssl_parse_serverhello_tlsext:serverhello tlsext"));
+        
+        String serverMessage = serverCauseRef.get().getMessage();
+        assertTrue(serverMessage, serverMessage.contains("tlsv1 bad certificate status response"));
+    }
+
+    @Test(timeout = 10000L)
+    public void testClientExceptionOpenSsl() throws Exception {
+        testClientException(SslProvider.OPENSSL);
+    }
+
+    @Test(timeout = 10000L)
+    public void testClientExceptionOpenSslRefCnt() throws Exception {
+        testClientException(SslProvider.OPENSSL_REFCNT);
+    }
+
+    /**
+     * Testing what happens if the {@link ClientOcspContext} throws an {@link Exception}.
+     * 
+     * The exception should bubble up on the client side and the connection should get closed.
+     */
+    private void testClientException(SslProvider sslProvider) throws Exception {
```

Comment:
This test and `testServerException` produce the NPE described in https://github.com/netty/netty/issues/6336.

OpenSSL calls the `verify()` method (via the OCSP callback - see PR in netty-tcnative) and this test throws purposely an IllegalStateException to see what happens.

I'm expecting a DecoderException to bubble up (with the IllegalStateException as cause) but instead I'm getting DecoderException that was caused by a NPE.